apps/builder/app/shared/tree-utils.ts

```diff
@@ -15,6 +15,49 @@ import {
 } from "@webstudio-is/project-build";
 import { getComponentMeta } from "@webstudio-is/react-sdk";
 
+// slots can have multiple parents so instance should be addressed
+// with full rendered path to avoid double selections with slots
+// and support deletion of slot child from specific parent
+// address starts with target instance and ends with root
+export type InstanceAddress = Instance["id"][];
+
+// provide an address starting with ancestor id
+// useful to select parent instance or one of breadcrumbs instances
+export const getAncestorInstanceAddress = (
+  instanceAddress: InstanceAddress,
+  ancestorId: Instance["id"]
+): undefined | InstanceAddress => {
+  const ancestorIndex = instanceAddress.indexOf(ancestorId);
+  if (ancestorIndex === -1) {
+    return undefined;
+  }
+  return instanceAddress.slice(ancestorIndex);
+};
+
+// this utility is temporary solution to compute instance addresses
+// before all logic is migrated to get it from rendered context
+// @todo should be deleted before adding slots
+export const getInstanceAddress = (
+  instances: Instances,
+  instanceId: Instance["id"]
+) => {
+  const parentInstancesById = new Map<Instance["id"], Instance["id"]>();
+  for (const instance of instances.values()) {
```

Comment:
is this iterating over ALL instances of the project?

Reply:
Yep, will get rid of it in upcoming PRs